hyperparameter borrowing and software stealing rules

johntran-nv opened the pull request.

christ1ne commented:
Can we add one line saying the borrowing can happen at most once to prevent the infinite loop of borrowing each others HP between 2 submitters? Is this a real concern?

johntran-nv commented on training_rules.adoc:
```diff
@@ -352,7 +352,14 @@ OPEN: Hyperparameters and optimizer may be freely changed.
+
+During the review period as described in the Submission Rules, a submitter may replace the hyperparameters in their implementation of a benchmark with hyperparameters from another submitter's implementation of the same benchmark. By default, they may change batch size (local batch size, global batch size, batchnorm span), but must replace all other hyperparameters as a group.
+
+With evidence that the resulting model, using the same batch size as the other submitter's implementation, converges worse in terms of epochs required, the submitter may make a minimum number of additional hyperparameter changes for the purpose of improving convergence and achieving comparable, but not better, convergence in epochs compared to the other submitter's implementation, but preserving  ...
```
Consider removing paragraph at line 359.

johntran-nv replied:
After discussing internally, we think that this paragraph should remain, because this is specifically talking about convergence between two submissions, which is distinctly different than the convergence rules we've come up with for v1.0, which compare a submission to the reference.

johntran-nv replied:
Added some more language around treating the "other" submission as the reference for this comparison sake.

johntran-nv commented:
@petermattson I believe we did discuss in the Training WG.  How about I send it out to the Training email alias to give folks another chance to chime in, if they want, and if no one objects in a couple of days, we can merge?

petermattson commented:
Sounds good.

On Fri, Feb 26, 2021 at 2:22 PM johntran-nv <notifications@github.com>
wrote:

> @petermattson <https://github.com/petermattson> I believe we did discuss
> in the Training WG. How about I send it out to the Training email alias to
> give folks another chance to chime in, if they want, and if no one objects
> in a couple of days, we can merge?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mlcommons/training_policies/pull/416#issuecomment-786924945>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIIVUHPZM5W3FWR3Y3UJ2T3TBANLTANCNFSM4WCVEVDQ>
> .
>

christ1ne commented on training_rules.adoc:
```diff
@@ -397,6 +404,10 @@ Additional exemptions need to be explicitly requested and approved in advance. I
 
+== Software Borrowing ==
+
+If a system software (framework, runtime, library, container, etc.) provider submits using software version that was released fewer than 14 days prior to the submission deadline, then other submitters using the same software are allowed to resubmit after updating their system software before the resubmission deadline during review period. Software borrowing applies only to system software, only t ...
```
"before the resubmission deadline during review period. " is this too late for a proper review? How about 1 week after the submission deadline?

fjtabaru replied:
Thank you.
I think some participants will be saved by this rule, as we were.
I agree that the entire review period is too long.
I think one week to two week is good.

christ1ne replied:
Benchmark implementations should be **borrowed** as a whole only if the software provider’s submission introduces new APIs. -> shall we replace borrowed with updated?

tjablin replied:
I think almost every new software version introduces new APIs. Can a whole implementation be borrowed even if the newly introduced API is unused?

tjablin replied:
Intel is a system provider. If they submit using a new Linux kernel than me, and the kernel is less than 14 days old, and the new kernel introduces a new syscall, can I borrow their entire benchmark implementation? Does their need to be a logical connection between the submitter and the recently updated piece of software?

johntran-nv replied:
@christ1ne , I think "borrowed" is more appropriate because it emphasizes that you have to take it as a whole.  Updated implies you don't have to take it as a whole, doesn't it?

@tjablin , I think new APIs are not a problem, because the implementation wouldn't call those new APIs, and the software would still work, right?  Or am I misunderstanding your point here?

DilipSequeira replied:
I agree we shouldn't have any ambiguity about when and how a component may be updated using these rules - I'd just rather not use a somewhat abstruse meaning of "borrow" to capture the concept.

DilipSequeira replied:
3 suggestions:
* move the SW rules to the main policies document so that they're uniform across WGs.
* use "adoption" instead of "borrowing" for software.
* expand the criteria for adoption along the following lines:

_For a given round of MLPerf, the "canonical version" of a software component shall be defined as the public version as of 14 days before submission. If the software is open source the canonical version shall be the one compiled with the default compilation options. If a system software provider submits with a component whose version is other than the canonical version, other submitters using the same component are allowed to update their submission to use that version._

This covers two issues that arose in Inference 1.0: a software owner using an older version where there was a regression in the newer version, and a software owner compiling with other than the default settings.

johntran-nv replied:
With my latest update, I believe I've addressed all of this feedback, with 2 exceptions.  

1. I'm just updating training policies for now.  We can move it to the main policies doc later, if all other WG's agree.
2. exact dates are still not defined concretely here, but we are currently working on nailing down the exact dates in the Training WG, so we'll incorporate that into the docs in the right places when done.

With those, I think we can go ahead and merge into the training_policies doc.

christ1ne commented on training_rules.adoc:
```diff
@@ -352,7 +352,14 @@ OPEN: Hyperparameters and optimizer may be freely changed.
-During the review period as described in the Submission Rules, a submitter may replace the hyperparameters in their implementation of a benchmark with hyperparameters from another submitter's implementation of the same benchmark. By default, they may or may not replace batch size but must replace all other hyperparameters as a group. With evidence that the resulting model converges worse in terms ...
+Submitters are expected to use their best efforts to submit with optimal hyperparameters for their system.  The intent of Hyperparameter Borrowing is to allow a submitter to update their submission to reflect what they would have submitted had they known about more optimal hyperparameters before submitting, without knowing any other info (ie the performance of other submissions).
+
+During the review period as described in the Submission Rules, a submitter may replace the hyperparameters in their implementation of a benchmark with hyperparameters from another submitter's implementation of the same benchmark. By default, they may change batch size (local batch size, global batch size, batchnorm span), but must replace all other hyperparameters as a group.
```
a submitter may replace the hyperparameters **once per benchmark entry** in their implementation of a benchmark with hyperparameters from another submitter's implementation of the same benchmark.

It maybe better to limit to one time borrowing

johntran-nv commented:
@christ1ne (or anyone else who wants), I believe I've addressed all the feedback, but want another set of eyes on it before I merge.  Could you please re-review?

DilipSequeira commented on training_rules.adoc:
```diff
@@ -397,6 +404,10 @@ Additional exemptions need to be explicitly requested and approved in advance. I
 
+== Software Adoption ==
+
+For a given round of MLPerf, the "canonical version" of a software component shall be defined as the public version as of 14 days before submission. If the software is open source, the canonical version shall be the one compiled with the default compilation options. If a system software provider submits with a component whose version is other than the canonical version, and uses software that was ...
```
I think we need to delete ", and uses software that was released fewer than 14 days prior to the submission deadline," since otherwise this doesn't cover the case seen in inference 1.0, where a submitter submitted with a year-old version of the SW because the current version had a regression. The "canonical version" here is the one that everyone would normally expect to use - and if the provider isn't using it, *for whatever reason*, other submitters should have the chance to update their submissions.

christ1ne replied:
agree with Dilip here about the deletion since you can borrow new version or the old version if there is unexpected regression on the canonical version here.

maybe we can clarify in the FAQ somewhere if two submitters use the same version but different compilation options, can they borrow from each other. It is not really that clear from the text.

@drcanchi

johntran-nv replied:
deleted

johntran-nv commented:
@christ1ne , @DilipSequeira , are we good to merge now?

DilipSequeira commented:
lgtm

christ1ne commented:
LGTM
